src/ir/auth_logic/ast_construction.cc

```diff
@@ -90,14 +102,53 @@ static Fact ConstructFact(AuthLogicParser::FactContext& context) {
   return ConstructFact(*CHECK_NOTNULL(ctx.fact()));
 }
 
+static Assertion ConstructAssertion(
+    AuthLogicParser::AssertionContext& context) {
+  if (auto* ctx =
+          dynamic_cast<AuthLogicParser::FactAssertionContext*>(&context)) {
+    return Assertion(ConstructFact(*CHECK_NOTNULL(ctx->fact())));
+  }
+  auto& ctx = *CHECK_NOTNULL(
+      dynamic_cast<AuthLogicParser::HornClauseAssertionContext*>(&context));
+  std::vector<BaseFact> base_facts = utils::MapIter<BaseFact>(
+      ctx.rvalue(), [](AuthLogicParser::RvalueContext* rvalue_context) {
+        return ConstructRvalue(*CHECK_NOTNULL(rvalue_context));
+      });
+  return Assertion(ConditionalAssertion(
+      ConstructFact(*CHECK_NOTNULL(ctx.fact())), std::move(base_facts)));
+}
+
+static SaysAssertion ConstructSaysAssertion(
+    AuthLogicParser::SaysAssertionContext& context) {
+  if (auto* ctx = dynamic_cast<AuthLogicParser::SaysSingleContext*>(&context)) {
+    std::vector<Assertion> assertions;
+    assertions.push_back(ConstructAssertion(*CHECK_NOTNULL(ctx->assertion())));
+    return SaysAssertion(ConstructPrincipal(*CHECK_NOTNULL(ctx->principal())),
+                         std::move(assertions));
```

Comment:
Optional, but the following would also be OK:
```suggestion
    return SaysAssertion(ConstructPrincipal(*CHECK_NOTNULL(ctx->principal())),
                         {ConstructAssertion(*CHECK_NOTNULL(ctx->assertion()))});
```